RecFCCeeCalorimeter/src/components/CalibrateCaloClusters.cpp

```diff
@@ -319,43 +351,70 @@ StatusCode CalibrateCaloClusters::calibrateClusters(const edm4hep::ClusterCollec
-  std::fill(energiesInLayer.begin(), energiesInLayer.end(), 0.0);
+  std::fill(energiesInLayers.begin(), energiesInLayers.end(), 0.0);
+
+  // get total cluster energy
+  double ecl = cluster.getEnergy();
 
-  // loop over the various readouts/subsystems/...
-  int startPositionToFill = 0;
-  for (size_t k = 0; k < m_readoutNames.size(); k++)
+  // if all inputs are available as shapeParameters, use them
+  if ((int) m_inputPositionsInShapeParameters.size() == m_numLayersTotal)
   {
-    if (k > 0)
+    // the shapeParameters already contain energy fractions so we
+    // do not divide by cluster energy
+    for (int i=0; i < (int) m_numLayersTotal; i++)
     {
-      startPositionToFill += m_numLayers[k - 1];
+      energiesInLayers[i] = cluster.getShapeParameters(m_inputPositionsInShapeParameters[i]);
     }
-    int systemID = m_systemIDs[k];
-    int firstLayer = m_firstLayerIDs[k];
-    std::string layerField = m_layerFieldNames[k];
-    std::string readoutName = m_readoutNames[k];
-    dd4hep::DDSegmentation::BitFieldCoder *decoder = m_geoSvc->getDetector()->readout(readoutName).idSpec().decoder();
-
-    // for each readout, loop over the cells and calculate the energies in the layers
-    // of that subsystem
-    for (auto cell = cluster.hits_begin(); cell != cluster.hits_end(); ++cell)
+    // add as last input the total raw cluster energy
+    energiesInLayers[m_numLayersTotal] = ecl;
+  }
+  else
+  {
+    // calculate the energy fractions from the cells
+    // loop over the various readouts/subsystems/...
+    int startPositionToFill = 0;
+    for (int k = 0; k < (int) m_readoutNames.size(); k++)
     {
-      dd4hep::DDSegmentation::CellID cellID = cell->getCellID();
-      if (decoder->get(cellID, "system") != systemID)
+      if (k > 0)
       {
-        continue;
+        startPositionToFill += m_numLayers[k - 1];
       }
-      int layer = decoder->get(cellID, layerField);
-      energiesInLayer[startPositionToFill + layer - firstLayer] += cell->getEnergy();
+      int systemID = m_systemIDs[k];
+      int firstLayer = m_firstLayerIDs[k];
+      std::string layerField = m_layerFieldNames[k];
+      std::string readoutName = m_readoutNames[k];
+      dd4hep::DDSegmentation::BitFieldCoder *decoder = m_geoSvc->getDetector()->readout(readoutName).idSpec().decoder();
+
+      // for each readout, loop over the cells and calculate the energies in the layers
+      // of that subsystem
+      for (auto cell = cluster.hits_begin(); cell != cluster.hits_end(); ++cell)
+      {
+        dd4hep::DDSegmentation::CellID cellID = cell->getCellID();
+        if (decoder->get(cellID, "system") != systemID)
+        {
+          continue;
+        }
+        int layer = decoder->get(cellID, layerField);
+        energiesInLayers[startPositionToFill + layer - firstLayer] += cell->getEnergy();
+      }
+    }
+    // divide by the cluster energy to prepare the inputs for the MVA
+    for (int k = 0; (int) k < m_numLayersTotal; ++k)
```

Comment:
Maybe `unsigned int`? as in many other places.

Reply:
changed to unsigned